flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java

```diff
@@ -228,91 +211,111 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
 	 * @return collection of partitions which still occupy some resources locally on this task executor
 	 * and have been not released yet.
 	 */
-	public Collection<ResultPartitionID> getUnreleasedPartitions() {
+	@Override
+	public Collection<ResultPartitionID> getPartitionsOccupyingLocalResources() {
 		return resultPartitionManager.getUnreleasedPartitions();
 	}
 
 	// --------------------------------------------------------------------------------------------
 	//  Create Output Writers and Input Readers
 	// --------------------------------------------------------------------------------------------
 
-	public ResultPartition[] createResultPartitionWriters(
-			String taskName,
-			ExecutionAttemptID executionId,
-			Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
-			MetricGroup outputGroup,
-			MetricGroup buffersGroup) {
+	@Override
+	public ShuffleIOOwnerContext createShuffleIOOwnerContext(
+			String ownerName,
+			ExecutionAttemptID executionAttemptID,
+			MetricGroup parentGroup) {
+		MetricGroup nettyGroup = createShuffleIOOwnerMetricGroup(checkNotNull(parentGroup));
+		return new ShuffleIOOwnerContext(
+			checkNotNull(ownerName),
+			checkNotNull(executionAttemptID),
+			parentGroup,
+			nettyGroup.addGroup(METRIC_GROUP_INPUT),
```

Comment:
It might be better to put these two `addGroup` covered by above `createShuffleIOOwnerMetricGroup` then return tuple2 to reference here because all the metrics related operations should be done inside `NettyShuffleMetricFactory`

Reply:
I thought about it, I think the operation is too primitive atm for an extra utility method. Besides, I think that in general `tuple2` is very adhoc and not so readable.

Reply:
Actually I also do not like the way of tuple. 
But we provide the method of `createShuffleIOOwnerMetricGroup` in factory, and it is only used for adding group simple. So it might bring confused that which `addGroup` should be done inside factory and which should be done outside. The scope/rule seems not clear.